src/librbd/AioRequest.cc

```diff
@@ -52,37 +57,160 @@ namespace librbd {
+      m_state = LIBRBD_AIO_READ_GUARD;
+    }
+  }
+
   bool AioRead::should_complete(int r)
   {
-    ldout(m_ictx->cct, 20) << "should_complete " << this << " " << m_oid << " " << m_object_off << "~" << m_object_len
-			   << " r = " << r << dendl;
+    ldout(m_ictx->cct, 20) << "AioRead::should_complete " << this
+                           << " " << m_oid << " " << m_object_off
+                           << "~" << m_object_len << " r = " << r
+                           << " cor = " << is_cor(m_ictx, m_snap_id)
+                           << " readonly = " << m_ictx->read_only
+                           << dendl;
 
-    if (!m_tried_parent && r == -ENOENT) {
-      RWLock::RLocker l(m_ictx->snap_lock);
-      RWLock::RLocker l2(m_ictx->parent_lock);
+    bool finished = true;
 
-      // calculate reverse mapping onto the image
-      vector<pair<uint64_t,uint64_t> > image_extents;
-      Striper::extent_to_file(m_ictx->cct, &m_ictx->layout,
-			    m_object_no, m_object_off, m_object_len,
-			    image_extents);
+    switch (m_state) {
+    case LIBRBD_AIO_READ_GUARD:
+      ldout(m_ictx->cct, 20) << "should_complete " << this
+                             << " READ_CHECK_GUARD" << dendl;
+
+      // This is the step to read from parent
+      if (!m_tried_parent && r == -ENOENT) {
+        RWLock::RLocker l(m_ictx->snap_lock);
+        RWLock::RLocker l2(m_ictx->parent_lock);
+
+        // calculate reverse mapping onto the image
+        vector<pair<uint64_t,uint64_t> > image_extents;
+        Striper::extent_to_file(m_ictx->cct, &m_ictx->layout,
+			        m_object_no, m_object_off, m_object_len,
+			        image_extents);
+
+        uint64_t image_overlap = 0;
+        r = m_ictx->get_parent_overlap(m_snap_id, &image_overlap);
+        if (r < 0) {
+          assert(0 == "FIXME");
+        }
+        uint64_t object_overlap = m_ictx->prune_parent_extents(image_extents, image_overlap);
+        if (object_overlap) {
+          m_tried_parent = true;
+
+          if (is_cor(m_ictx, m_snap_id)) {
+            // If there is a valid object being coyping up, directly extract
+            // content and finish up.
+            Mutex::Locker l3(m_ictx->copyup_list_lock);
+
+            map<uint64_t, CopyupRequest*>::iterator it =
+              m_ictx->copyup_list.find(m_object_no);
+            if (it != m_ictx->copyup_list.end()) {
+              Mutex::Locker l4(it->second->get_lock());
+              if (it->second->is_ready()) {
+                ceph::bufferlist &copyup_data = it->second->get_copyup_data();
+                m_read_data.substr_of(copyup_data, m_object_off, m_object_len);
+                ldout(m_ictx->cct, 20) << __func__ << " extract content from copyup_list, obj-"
+                                       << m_object_no << dendl;
+                finished = true;
+                break;
+              }
+            }
+          }
+
+          read_from_parent(image_extents);
```

Comment:
It's possible for ```read_from_parent``` to instantly call back into this instance's ```should_complete``` method.  If that happens, ```finished``` will be flagged as true and this instance will be deleted -- and then control will return here within a deleted object.